### PR TITLE
Persistence: Support for idempotency in ingestion and Write Statistics

### PR DESCRIPTION
**Pull request by prasar-ashutosh**
#### What type of PR is this?
- Improvement

#### What does this PR do / why is it needed ?
1. Idempotency support in ingestion
2. Write ingest batch statistics in batch metadata table
3. Introduce batch ingest request as a separate column in batch metadata table
4. Tests

#### Does this PR introduce a user-facing change?
NO

### CODE REVIEW

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java [hunk 1]
```diff
@@ -104,6 +104,12 @@ default boolean collectStatistics()
             return false;
         }
 
+        @Default
+        default boolean writeStatistics()
+        {
+            return false;
```
Comment (rengam32) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java:110]: Curious why the default is false?

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java [hunk 2]
```diff
@@ -540,6 +572,43 @@ private void acquireLock()
...
+            {
+                List<Map<String, Object>> metadataResults = tabularDataList.get(0).getData();
+                for (Map<String, Object> metadata: metadataResults)
+                {
+                    Timestamp ingestionTimestampUTC = (Timestamp) metadata.get(metadataDataset.batchStartTimeField());
+                    String batchStatus = (String) metadata.get(metadataDataset.batchStatusField());
+                    batchStatus = batchStatus.equalsIgnoreCase(batchSuccessStatusValue())  ? IngestStatus.SUCCEEDED.name() : batchStatus;
+                    IngestorResult ingestorResult = IngestorResult.builder()
+                            .batchId((int) metadata.get(metadataDataset.tableBatchIdField()))
+                            .putAllStatisticByName(readValueAsMap((String) metadata.get(metadataDataset.batchStatisticsField())))
```
Comment (kumuwu) [inline legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java:598]: We may have to do a String.valueOf(), as this casting may cause error in DuckDB and Postgres.
In the Postgres and DuckDB PR, we have made a such a change in a similar place: https://github.com/finos/legend-engine/blob/293388680d5088f37d31a04c3dcb52461f628d68/legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/ApiUtils.java#L232

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java (hunk 1 above)
Comment (prasar-ashutosh) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/planner/Planner.java:110]: This is for back ward compatibility with CNAS

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/logicalplan/LogicalPlanFactory.java [hunk 3]
```diff
@@ -99,6 +102,22 @@ public static LogicalPlan getLogicalPlanForNextBatchId(Datasets datasets)
         return LogicalPlan.builder().addOps(selection).build();
     }
 
+    public static LogicalPlan getLogicalPlanForIdempotencyCheck(MetadataDataset metadataDataset, String ingestRequestId, String mainDatasetName)
```
[the remark this answers is not in the document]
Comment (prasar-ashutosh) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-logical-plan/src/main/java/org/finos/legend/engine/persistence/components/logicalplan/LogicalPlanFactory.java]: Done

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java [hunk 4]
```diff
@@ -540,6 +572,43 @@ private void acquireLock()
...
+            List<TabularData> tabularDataList = executor.executePhysicalPlanAndGetResults(physicalPlan);
+            MetadataDataset metadataDataset = enrichedDatasets.metadataDataset().get();
+            if (!tabularDataList.isEmpty())
+            {
+                List<Map<String, Object>> metadataResults = tabularDataList.get(0).getData();
+                for (Map<String, Object> metadata: metadataResults)
+                {
+                    Timestamp ingestionTimestampUTC = (Timestamp) metadata.get(metadataDataset.batchStartTimeField());
+                    String batchStatus = (String) metadata.get(metadataDataset.batchStatusField());
+                    batchStatus = batchStatus.equalsIgnoreCase(batchSuccessStatusValue())  ? IngestStatus.SUCCEEDED.name() : batchStatus;
```
[the remark this answers is not in the document]
Comment (prasar-ashutosh) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java:595]: This is to map "DONE" to enum SUCCEEDED.

File: legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java (hunk 2 above)
Comment (prasar-ashutosh) [reply legend-engine-xts-persistence/legend-engine-xt-persistence-component/legend-engine-xt-persistence-component-relational-core/src/main/java/org/finos/legend/engine/persistence/components/relational/api/RelationalIngestorAbstract.java]: Done

**Review by epsstan — APPROVED** [review]

**Pull request merged**